numpy/core/src/multiarray/iterators.c

```diff
@@ -640,35 +641,36 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
         obj = ind;
     }
 
-    if (PyArray_Check(obj)) {
-        /* Check for Boolean object */
-        if (PyArray_TYPE((PyArrayObject *)obj) == NPY_BOOL) {
-            ret = iter_subscript_Bool(self, (PyArrayObject *)obj);
-            Py_DECREF(indtype);
-        }
-        /* Check for integer array */
-        else if (PyArray_ISINTEGER((PyArrayObject *)obj)) {
-            PyObject *new;
-            new = PyArray_FromAny(obj, indtype, 0, 0,
-                              NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
-            if (new == NULL) {
-                goto fail;
-            }
-            Py_DECREF(obj);
-            obj = new;
-            new = iter_subscript_int(self, (PyArrayObject *)obj);
-            Py_DECREF(obj);
-            return new;
-        }
-        else {
-            goto fail;
-        }
+    /* Any remaining valid input is an array or has been turned into one */
+    if (!PyArray_Check(obj)) {
+        goto fail;
+    }
+
+    /* Check for Boolean array */
+    if (PyArray_TYPE((PyArrayObject *)obj) == NPY_BOOL) {
+        ret = iter_subscript_Bool(self, (PyArrayObject *)obj);
+        Py_DECREF(indtype);
         Py_DECREF(obj);
         return (PyObject *)ret;
     }
-    else {
-        Py_DECREF(indtype);
+
+    /* Only integer arrays left */
+    if (!PyArray_ISINTEGER((PyArrayObject *)obj)) {
+        goto fail;
+    }
+
+    Py_INCREF(indtype);
+    new = PyArray_FromAny(obj, indtype, 0, 0,
+                      NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
+    if (new == NULL) {
+        goto fail;
     }
+    Py_DECREF(indtype);
+    Py_DECREF(obj);
+    obj = new;
+    new = iter_subscript_int(self, (PyArrayObject *)obj);
+    Py_DECREF(obj);
```

Comment:
These three lines can be spelt `Py_SETREF(new, iter_subscript_int(self, (PyArrayObject *)new))`

Reply:
It may make sense to use `ret` here instead of `new`, so that one returns `ret`. (Clearly we've gotten to the stage of totla nitpicking!)

Reply:
I'm OK with improving the code as much as possible while we're at it. Plus it's not my original code we're nitpicking here ;) I'll take both of those suggestions, thanks.

Reply:
@mhvk I've held back the renaming for now, because beyond the name `ret` is a `PyArrayObject *` and I'm not sure how I could cast safely. Furthermore, we're going to return a `PyObject *`, so it feels a bit weird to cast back and forth in two subsequent lines.

On a related note, is it not weird that some of the functions in `iterators.c` return `PyObject *`s while others `PyArrayObject *`s? In particular, one out of three `iter_subscript_*` functions returns the latter. If there's a sane way to use `ret` after all we can change it of course, I just don't know how.

Reply:
@adeak - definitely fine to keep as is. I notice that somewhere about `ret` is used in exactly the poor way you describe - but let's leave that out for now!